contracts/limitOrder/LimitOrderBook.sol

```diff
@@ -0,0 +1,161 @@
+        clearingHouse = clearingHouseArg;
+        accountBalance = IClearingHouse(clearingHouse).getAccountBalance();
+
+        emit ClearingHouseChanged(clearingHouseArg);
+    }
+
+    function setLimitOrderFeeVault(address limitOrderFeeVaultArg) external onlyOwner {
+        // LOB_LOFVINC: LimitOrderFeeVault Is Not a Contract
+        require(limitOrderFeeVaultArg.isContract(), "LOB_LOFVINC");
+        limitOrderFeeVault = limitOrderFeeVaultArg;
+
+        emit LimitOrderFeeVaultChanged(limitOrderFeeVaultArg);
+    }
+
+    /// @inheritdoc ILimitOrderBook
+    function fillLimitOrder(LimitOrder memory order, bytes memory signature) external override nonReentrant {
+        bytes32 orderHash = getOrderHash(order);
+        verifySigner(order, signature);
+
+        // LOB_OIFA: Order is filled already
+        require(_ordersStatus[orderHash] != ILimitOrderBook.OrderStatus.Filled, "LOB_OIFA");
+        // LOB_OIC: Order is cancelled
+        require(_ordersStatus[orderHash] != ILimitOrderBook.OrderStatus.Cancelled, "LOB_OIC");
+
+        int256 oldTakerPositionSize = IAccountBalance(accountBalance).getTakerPositionSize(
+            order.trader,
+            order.baseToken
+        );
+
+        (uint256 base, uint256 quote) = IClearingHouse(clearingHouse).openPositionFor(
+            order.trader,
+            IClearingHouse.OpenPositionParams({
+                baseToken: order.baseToken,
+                isBaseToQuote: order.isBaseToQuote,
+                isExactInput: order.isExactInput,
+                amount: order.amount,
+                oppositeAmountBound: order.oppositeAmountBound,
+                deadline: order.deadline,
+                sqrtPriceLimitX96: 0,
+                referralCode: ""
+            })
+        );
+
+        int256 newTakerPositionSize = IAccountBalance(accountBalance).getTakerPositionSize(
+            order.trader,
+            order.baseToken
+        );
+
+        if (order.reduceOnly) {
+            // LOB_TINR : this it not reduceOnly
+            require(
+                oldTakerPositionSize.mul(newTakerPositionSize) > 0 &&
+                    oldTakerPositionSize.abs() > newTakerPositionSize.abs(),
+                "LOB_TINR"
+            );
+        }
+
+        _ordersStatus[orderHash] = ILimitOrderBook.OrderStatus.Filled;
+
+        address keeper = _msgSender();
+        uint256 keeperFee = ILimitOrderFeeVault(limitOrderFeeVault).disburse(keeper, quote);
+
+        emit LimitOrderFilled(order.trader, order.baseToken, orderHash, keeper, keeperFee);
+    }
+
+    /// @inheritdoc ILimitOrderBook
+    function cancelLimitOrder(LimitOrder memory order) external override {
+        // LOB_OSMBS: Order's Signer Must Be Sender
+        require(_msgSender() == order.trader, "LOB_OSMBS");
+
+        // we didn't require `signature` as input like fillLimitOrder(),
+        // so trader can actually cancel an order that is not existed
+        bytes32 orderHash = getOrderHash(order);
```

Comment:
I'm not sure about this, should we change to `cancelLimitOrder(LimitOrder memory order, bytes memory signature)`?

Reply:
In the first place, I agree with you but after thinking a while, I think we should keep what we have now. The reason why is traders will submit their limit orders more carefully bcs they need to pay gas to cancel it. Might mitigate our keeper being attacked by spam orders.